File: frontend/src/types/smart-search.ts
/**
 * Smart Search Domain Models
 * 
 * Core business objects for the Smart Search feature.
 * These are shared data structures used across multiple components.
 * API-specific request/response models are defined in @/lib/api/smartSearchApi.
 */

import { CanonicalResearchArticle } from './canonical_types';

// Re-export the canonical type for convenience
export type { CanonicalResearchArticle } from './canonical_types';

// Legacy alias for backward compatibility during migration  
export type SearchArticle = CanonicalResearchArticle;

export interface SearchPaginationInfo {
  total_available: number;
  returned: number;
  offset: number;
  has_more: boolean;
}

export interface FilteredArticle {
  article: CanonicalResearchArticle;
  passed: boolean;
  confidence: number;
  reasoning: string;
}

export interface FilteringProgress {
  total: number;
  processed: number;
  accepted: number;
  rejected: number;
  current_article?: string;
}

// More flexible version for session storage
export interface FilteredArticleForSession {
  article: Record<string, any>; // Flexible schema for session storage
  passed: boolean;
  confidence: number;
  reasoning: string;
}

// ============================================================================
// Session Management Types (matching backend schemas)
// ============================================================================

export interface SmartSearchSession {
  id: string;
  user_id: string;
  created_at: string | null;
  updated_at: string | null;
  original_question: string;
  generated_evidence_spec: string | null;
  submitted_evidence_spec: string | null;
  generated_search_keywords: string | null;
  submitted_search_keywords: string | null;
  search_metadata: {
    total_available?: number;
    total_retrieved?: number;
    sources_searched?: string[];
    [key: string]: any;
  } | null;
  articles_retrieved_count: number;
  articles_selected_count: number;
  generated_discriminator: string | null;
  submitted_discriminator: string | null;
  filter_strictness: 'low' | 'medium' | 'high' | null;
  filtering_metadata: {
    accepted?: number;
    rejected?: number;
    total_processed?: number;
    custom_columns?: any[];
    [key: string]: any;
  } | null;
  filtered_articles: FilteredArticleForSession[] | null;
  status: string;
  last_step_completed: string | null;
  session_duration_seconds: number | null;
  total_api_calls: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
  total_tokens: number;
}

export interface SessionListResponse {
  sessions: SmartSearchSession[];
  total: number;
}

export interface SessionResetResponse {
  message: string;
  session: SmartSearchSession;
}

File: backend/schemas/smart_search.py
"""
Smart Search Schemas

Core domain models for the smart search feature.
These are shared data structures used across multiple services.
API-specific request/response models are defined in the router.
"""

from typing import List, Optional
from pydantic import BaseModel, Field
from schemas.canonical_types import CanonicalResearchArticle


class SearchPaginationInfo(BaseModel):
    """Pagination information for search results"""
    total_available: int = Field(..., description="Total number of results available")
    returned: int = Field(..., description="Number of results returned in this request")
    offset: int = Field(..., description="Number of results skipped")
    has_more: bool = Field(..., description="Whether there are more results available")


class FilteredArticle(BaseModel):
    """Article with filtering results - core domain model"""
    article: CanonicalResearchArticle
    passed: bool = Field(..., description="Whether article passed the filter")
    confidence: float = Field(..., description="Confidence score 0-1")
    reasoning: str = Field(..., description="Brief explanation of decision")


class FilteringProgress(BaseModel):
    """Progress update for filtering operations"""
    total: int
    processed: int
    accepted: int
    rejected: int
    current_article: Optional[str] = None


class SearchServiceResult(BaseModel):
    """Result from search service methods"""
    articles: List[CanonicalResearchArticle]
    pagination: SearchPaginationInfo
    sources_searched: List[str]


class OptimizedKeywordsResult(BaseModel):
    """Result from generate_optimized_search_keywords service method"""
    initial_keywords: str  # was: initial_query
    initial_count: int
    final_keywords: str    # was: final_query
    final_count: int
    refinement_description: str
    status: str  # 'optimal' | 'refined' | 'manual_needed'